### PR TITLE
Refactor and Divide some function

### PR DESCRIPTION
**Pull request by Ikuyadeu**
17592bd is just rename and add new line for lintr.
26dac1b is divide some function by messages kind(not important and reset is ok).

### CODE REVIEW

File: R/document.R [hunk 1]
```diff
@@ -0,0 +1,107 @@
+workspace <- R6::R6Class("textDocument",
```
Comment (randy3k) [inline R/document.R]: why do we need these refactoring? Are you planning to move all the handlers here?

File: R/server.R [hunk 2]
```diff
@@ -2,11 +2,11 @@
 run <- function(debug = FALSE, stdin = "stdin", stdout = "stdout") {
     logger$set_mode(debug = debug)
     langserver <- LanguageServer$new(stdout = stdout)
-    con <- file(stdin)
-    open(con, blocking = TRUE)
+    content <- file(stdin)
```
Comment (randy3k) [inline R/server.R]: thank for the cleaning here.

File: R/workspace.R [hunk 3]
```diff
@@ -0,0 +1,21 @@
+workspace <- R6::R6Class("workspace",
```
Comment (randy3k) [inline R/workspace.R]: again, why do you think a R6 class is necessary?

File: R/document.R (hunk 1 above)
Comment (Ikuyadeu) [reply R/document.R]: Right, I added class for identify `completionItem` and `textDocument`.
But dividing by file may be better than this?

File: R/workspace.R (hunk 3 above)
Comment (Ikuyadeu) [reply R/workspace.R]: Adding this class is for consistency `document.R` file, but don't need.

File: R/document.R (hunk 1 above)
Comment (randy3k) [reply R/document.R]: That's what I think. A R6 class does not seem necessary.

Comment (Ikuyadeu) [reply R/document.R]: Thanks, but I can't fix this now soon.
I can fix this here's morning(9~ hours ago)

Comment (randy3k) [reply R/document.R]: No problem.

PS: I think you meant 9 hr later.

**Comment by Ikuyadeu** [comment]
@randy3k Ok, fixed.

File: R/codeLens.R [hunk 4]
```diff
@@ -0,0 +1,3 @@
+resolve  <- function(self, id, params) {
```
Comment (randy3k) [inline R/codeLens.R]: we cannot just call it `resolve`, still need a prefix.

File: R/documentLink.R [hunk 5]
```diff
@@ -0,0 +1,3 @@
+resolve  <- function(self, id, params) {
```
Comment (randy3k) [inline R/documentLink.R]: same as another resolve

File: R/completionItem.R [hunk 6]
```diff
@@ -0,0 +1,3 @@
+resolve  <- function(self, id, params) {
```
Comment (randy3k) [inline R/completionItem.R]: same as another resolve

File: R/document.R [hunk 7]
```diff
@@ -0,0 +1,81 @@
+did_open <- function(self, params) {
```
Comment (randy3k) [inline R/document.R]: may be still a good idea to have the prefix document? like `document_did_open`?

File: R/document.R [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    textDocument <- params$textDocument
+    publish_diagnostics(self, textDocument$uri)
+}
+
+did_change <- function(self, params) {
```
Comment (randy3k) [inline R/document.R]: I saw you added comments for identify the functions as notification or request, may be we should also do the same thing here?

**Comment by Ikuyadeu** [comment]
Thank you, fixed.

File: R/document.R (hunk 7 above)
Comment (Ikuyadeu) [reply R/document.R]: I think not so, because we can identify from file name.

Comment (randy3k) [reply R/document.R]: But it is not how a R package works, all the functions are loaded to the same namespace no matter how the files are named. For instance, if there is another `did_open` in `workspace.R`. One of them will be overwritten by the other.

**Comment by randy3k** [comment]
You also need to change the name of the functions [here](https://github.com/REditorSupport/languageserver/blob/master/R/zzz.R#L9) and remove `notification_handlers` and `request_handler`.

File: R/document.R (hunk 7 above)
Comment (Ikuyadeu) [reply R/document.R]: I see, so I'll fix to `text_document_did_open`. (from https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md)
Is it OK?

File: R/general.R [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    # TODO: result lint result of the package
+    # lint_result <- lintr::lint_package(rootPath)
+}
+
+# Request
+on_shutdown <- function(self, id, params) {
+
+}
+
+# Request
```
Comment (randy3k) [inline R/general.R]: it is a notification.

File: R/general.R [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+on_exit <- function(self, params) {
+    self$will_exit <- TRUE
+}
+
+# Request
+cancel_request <- function(self, params) {
```
Comment (randy3k) [inline R/general.R]: it is also a notificaiton

File: R/general.R [hunk 11]
```diff
@@ -0,0 +1,31 @@
+on_initialize <- function(self, id, params) {
```
Comment (randy3k) [inline R/general.R]: `general.R` may be indeed too general. How about moving them to sub folders `handlers`?

Comment (Ikuyadeu) [reply R/general.R]: You mean this https://github.com/Ikuyadeu/languageserver/tree/refactor/R/handlers?

**Comment by Ikuyadeu** [comment]
more fix

**Comment by Ikuyadeu** [comment]
> general.R may be indeed too general. How about moving them to sub folders handlers?
 
You mean this https://github.com/Ikuyadeu/languageserver/tree/refactor/R/handlers?

**Comment by Ikuyadeu** [comment]
How is this folder

handlers/
  |--request/
  |              |--document.R
  |              |--workspace.R
  |     
  |--notification/
  |              |--document.R
  |              |--workspace.R

File: R/zzz.R [hunk 12]
```diff
@@ -1,11 +1,10 @@
-request_handlers <- list(
```
Comment (randy3k) [inline R/zzz.R]: You should not change `request_handlers` and `notification_handlers`, they are being dispatched at
https://github.com/REditorSupport/languageserver/blob/44050bbefe376ee66b2625a61e7954c49f5e1098/R/languageserver.R#L49-L72

**Comment by randy3k** [comment]
I think the current structure (`handlers/document.R`, `handlers/workspace.R`)... is fine

**Comment by Ikuyadeu** [comment]
OK, how is it?

**Comment by randy3k** [comment]
Sorry @Ikuyadeu 
I gave you wrong information, the R files have to be in the top level of the `R` directory.
Like this https://github.com/REditorSupport/languageserver/tree/master/R

**Comment by Ikuyadeu** [comment]
Oops, I'll be careful next.

**Comment by randy3k** [comment]
No problem. Thank you for the work.

**Pull request merged**